### PR TITLE
[WIP] [FG:InPlacePodVerticalScaling] Add /resize subresource to request pod resource resizing

### LINKED ISSUE #109553: [FG:InPlacePodVerticalScaling] Add /resize subresource to request pod resource resizing
### What would you like to be added?

In-place Pod vertical scaling [feature](/kubernetes/kubernetes/pull/102884/) currently relies on PATCH verb applied to Pod objects. This issue tracks a task to design and implement /resize subresource for Pods as opposed to PATCH, add RBAC controls to it, and additionally consider supporting it for Pod superset objects such as Jobs, RS, Deploys etc.

This came up during KEP discussions with Tim Hockin as we revisited first version of the API.
Please see the following discussion threads for context:
https://github.com/kubernetes/enhancements/pull/1883#discussion_r459156714
https://github.com/kubernetes/enhancements/pull/1883#discussion_r464627665
https://github.com/kubernetes/enhancements/pull/1883#discussion_r448752659
https://github.com/kubernetes/enhancements/pull/1883#discussion_r555668219
https://github.com/kubernetes/enhancements/pull/1883#issuecomment-814273899

### Why is this needed?

It may be cleaner and more desirable to use /resize subresource to request a desired resize, and such a resource can be optionally supported on objects such as Deployments, Jobs etc.

### PR DESCRIPTION
**Pull request by iholder101**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:
Fixes #109553

#### Does this PR introduce a user-facing change?

```release-note
Add /resize subresource to request pod resource resizing
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -6620,3 +6620,32 @@ type ImageVolumeSource struct {
 	// +optional
 	PullPolicy PullPolicy
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Resize is the query options to a Pod's resize call
+type Resize struct {
```
Comment (tallclair) [inline pkg/apis/core/types.go:6627]: A problem with these special-purpose request types is that admission controllers (and admission policy) only see this request type, and can't easily make decisions in the context of the full pod object. It also makes it more difficult to assert that there wasn't an intermediate change to the pod. In the KEP, I propose that the subresource type just be the full pod object. This is consistent with /status and /ephemeralContainers.

**Comment by tallclair** [comment]
In https://github.com/kubernetes/enhancements/pull/4704, I'm also proposing that we restrict the ability to modify resources to the subresource.

**Comment by tallclair** [comment]
With the KEP update merged, I think you can proceed with the implementation described in https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#subresource

In summary, the following changes should be made:
1. resize requests takes the full pod object as the request type, but only the resources & resize policy should be mutable
2. regular pod update should no longer allow resource mutation

We also need to figure out what we're doing in terms of managed fields. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/managedfields/scalehandler.go is how it's handled for the scale subresource, but it doesn't look like anything special is done for the pod subresources. cc @thockin @jpbetz 

/assign
/milestone v1.32
/triage accepted

**Comment by wendy-ha18** [comment]
Hi @tallclair @iholder101

Friendly reminder that code freeze is starting 02:00 UTC Friday November 8th 2024 (about 3 weeks from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Given this timeline and capacity, will this continue to be aimed for the 1.32 release?
Thanks! 😄

**Comment by jpbetz** [comment]
I don't want to further bog down this PR for alpha, but can we track in the KEP that quota admission is going to be bypassed by requests to a subresource that change resources and that we need to address that before this enhancement promotes to Beta (enabled-by-default)?

**Comment by tallclair** [comment]
This PR is superseded by https://github.com/kubernetes/kubernetes/pull/128266

/close

**Pull request closed without merging**